File: frontend/investment-dashboard/src/types/Investment.ts
export interface Investment {
  id: number;
  name: string;
  value: number;
  quantity: number;
  currency: string;
  date: string;
  description: string;
  category: string;
  accountId: number;
  account: {
    id: number;
    name: string;
    goal1?: number;
    goal2?: number;
    goal3?: number;
    goal4?: number;
    goal5?: number;
  };
  country: string;
  total: number;
}

export const Currency = {
  BRL: 'BRL',
  CAD: 'CAD',
  USD: 'USD'
} as const;

export const Category = {
  RendaFixa: 'RendaFixa',
  Stocks: 'Stocks',
  FIIs: 'FIIs',
  ETF: 'ETF',
  Bonds: 'Bonds',
  ManagedPortfolio: 'ManagedPortfolio',
  Cash: 'Cash',
  ManagedPortfolioBlock: 'ManagedPortfolioBlock'
} as const;

export type Currency = typeof Currency[keyof typeof Currency];
export type Category = typeof Category[keyof typeof Category];

// Helper functions to convert string values to display strings
export const getCurrencyLabel = (currency: string): string => {
  switch (currency) {
    case 'BRL': return 'BRL (Brazil)';
    case 'CAD': return 'CAD (Canada)';
    case 'USD': return 'USD (United States)';
    default: return 'Unknown';
  }
};

export const getCategoryLabel = (category: string): string => {
  switch (category) {
    case 'RendaFixa': return 'Renda Fixa';
    case 'Stocks': return 'Stocks';
    case 'FIIs': return 'FIIs';
    case 'ETF': return 'ETF';
    case 'Bonds': return 'Bonds';
    case 'ManagedPortfolio': return 'Managed Portfolio';
    case 'Cash': return 'Cash';
    case 'ManagedPortfolioBlock': return 'Managed Portfolio - block';
    default: return 'Unknown';
  }
};

export interface CreateInvestmentRequest {
  name: string;
  value: number;
  quantity: number;
  currency: string;
  date: string;
  description: string;
  category: string;
  accountId: number;
}

export interface GroupedInvestment {
  category: string;
  account: string;
  name: string;
  totalQuantity: number;
  averageValue: number;
  total: number;
  currency: string;
  country: string;
}

export interface AssetByCategory {
  category: string;
  total: number;
  count: number;
  percentage: number;
}

export interface InvestmentPerformance {
  investmentId: number;
  name: string;
  symbol: string;
  category: string;
  currency: string;
  quantity: number;
  purchasePrice: number;
  currentPrice?: number;
  totalInvested: number;
  currentValue: number;
  gainLoss: number;
  gainLossPercentage: number;
  lastUpdated: string;
  hasCurrentPrice: boolean;
}

export interface AccountPerformance {
  accountId: number;
  accountName: string;
  investments: InvestmentPerformance[];
  totalInvested: number;
  currentValue: number;
  totalGainLoss: number;
  totalGainLossPercentage: number;
}

export interface AccountGoalProgress {
  accountId: number;
  accountName: string;
  currentValue: number;
  currency: string;
  goals: {
    year1?: number;
    year2?: number;
    year3?: number;
    year4?: number;
    year5?: number;
  };
  performance?: AccountPerformance;
}

export interface DashboardData {
  allInvestments: Investment[];
  groupedInvestments: GroupedInvestment[];
  assetsByAccount: { account: string; total: number }[];
  assetsByCountry: { country: string; total: number }[];
  assetsByCategory: AssetByCategory[];
  accountGoals: AccountGoalProgress[];
  timelineData?: InvestmentTimelineData;
}

export interface InvestmentTimelineData {
  timelinePoints: TimelinePoint[];
  goalMarkers: GoalMarker[];
  currentTotalValue: number;
  currentBrlValue: number;
  currentCadValue: number;
  // Optional properties for original currency display mode
  originalCurrentBrlValue?: number;
  originalCurrentCadValue?: number;
  displayMode?: string;
}

export interface TimelinePoint {
  date: string;
  totalValue: number;
  brlValue: number;
  cadValue: number;
  // Optional properties for original currency display mode
  originalBrlValue?: number;
  originalCadValue?: number;
  displayMode?: string;
}

export interface GoalMarker {
  year: number;
  value: number;
  currency: string;
  accountName: string;
  label: string;
}

File: backend/FollowInvestments.Api/Models/Investment.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Text.Json.Serialization;

namespace FollowInvestments.Api.Models;

public class Investment
{
    [Key]
    [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
    public int Id { get; set; }

    [Required]
    [MaxLength(200)]
    public string Name { get; set; } = string.Empty;

    [Required]
    [Column(TypeName = "decimal(18,2)")]
    public decimal Value { get; set; }

    [Required]
    [Column(TypeName = "decimal(18,4)")]
    public decimal Quantity { get; set; }

    [Required]
    public Currency Currency { get; set; }

    [Required]
    public DateTime Date { get; set; }

    [Required]
    [MaxLength(500)]
    public string Description { get; set; } = string.Empty;

    [Required]
    public Category Category { get; set; }

    [Required]
    public int AccountId { get; set; }

    // Foreign key to User
    [Required]
    public int UserId { get; set; }

    // Navigation properties
    public Account Account { get; set; } = null!;
    public virtual User User { get; set; } = null!;

    public string Country => Currency == Currency.BRL ? "Brazil" : "Canada";
    
    public decimal Total => Quantity * Value;
}

public enum Currency
{
    BRL,
    CAD
}

public enum Category
{
    RendaFixa,
    Stocks,
    FIIs,
    ETF,
    Bonds,
    ManagedPortfolio,
    Cash,
    ManagedPortfolioBlock
}

public class CreateInvestmentRequest
{
    [Required]
    [MaxLength(200)]
    public string Name { get; set; } = string.Empty;

    [Required]
    public decimal Value { get; set; }

    [Required]
    public decimal Quantity { get; set; }

    [Required]
    public Currency Currency { get; set; }

    [Required]
    public DateTime Date { get; set; }

    [Required]
    [MaxLength(500)]
    public string Description { get; set; } = string.Empty;

    [Required]
    public Category Category { get; set; }

    [Required]
    public int AccountId { get; set; }
}

public class UpdateInvestmentRequest
{
    [JsonPropertyName("id")]
    public int Id { get; set; }

    [JsonPropertyName("name")]
    public string Name { get; set; } = string.Empty;

    [JsonPropertyName("value")]
    public decimal Value { get; set; }

    [JsonPropertyName("quantity")]
    public decimal Quantity { get; set; }

    [JsonPropertyName("currency")]
    public Currency Currency { get; set; }

    [JsonPropertyName("date")]
    public DateTime Date { get; set; }

    [JsonPropertyName("description")]
    public string Description { get; set; } = string.Empty;

    [JsonPropertyName("category")]
    public Category Category { get; set; }

    [JsonPropertyName("accountId")]
    public int AccountId { get; set; }
}